fpx-workers/src/data.rs

```diff
@@ -0,0 +1,144 @@
+            database: Arc::new(database),
+        }
+    }
+
+    async fn fetch_one<T>(&self, query: impl Into<String>, values: &[JsValue]) -> Result<T>
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        let prepared_statement = self.database.prepare(query);
+
+        let result = prepared_statement
+            .bind(values)
+            .map_err(|err| DbError::InternalError(err.to_string()))?; // TODO: Correct error
+
+        let result = result
+            .first(None)
+            .await
+            .map_err(|err| DbError::InternalError(err.to_string()))?
+            .ok_or(DbError::NotFound)?; // TODO: Correct error
+
+        Ok(result)
+    }
+
+    async fn fetch_all<T>(&self, query: impl Into<String>, values: &[JsValue]) -> Result<Vec<T>>
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        let prepared_statement = self.database.prepare(query);
+
+        let result = prepared_statement
+            .bind(values)
+            .map_err(|err| DbError::InternalError(err.to_string()))?; // TODO: Correct error
+
+        let result = result
+            .all()
+            .await
+            .map_err(|err| DbError::InternalError(err.to_string()))? // TODO: Correct error
+            .results()
+            .map_err(|err| DbError::InternalError(err.to_string()))?; // TODO: Correct error
+
+        Ok(result)
+    }
+}
+
+#[async_trait]
+impl Store for D1Store {
+    async fn start_readonly_transaction(&self) -> Result<Transaction> {
+        // Let's fake this for now
+        Ok(Transaction::default())
+    }
+    async fn start_readwrite_transaction(&self) -> Result<Transaction> {
+        // Let's fake this for now
+        Ok(Transaction::default())
+    }
+
+    async fn commit_transaction(&self, _tx: Transaction) -> Result<(), DbError> {
+        // Let's fake this for now
+        Ok(())
+    }
+    async fn rollback_transaction(&self, _tx: Transaction) -> Result<(), DbError> {
+        // Let's fake this for now
+        Ok(())
+    }
+
+    async fn span_get(
+        &self,
+        _tx: &Transaction,
+        trace_id: String,
+        span_id: String,
+    ) -> Result<models::Span> {
+        SendFuture::new(async {
+            self.fetch_one(
+                "SELECT * FROM spans WHERE trace_id=$1 AND span_id=$2",
+                &[trace_id.into(), span_id.into()],
+            )
+            .await
+        })
+        .await
+    }
+
+    async fn span_list_by_trace(
+        &self,
+        _tx: &Transaction,
+        trace_id: String,
+    ) -> Result<Vec<models::Span>> {
+        SendFuture::new(async {
+            self.fetch_all("SELECT * FROM spans WHERE trace_id=$1", &[trace_id.into()])
+                .await
+        })
+        .await
+    }
+
+    async fn span_create(
+        &self,
+        _tx: &Transaction,
+        span: models::Span,
+    ) -> Result<models::Span, DbError> {
+        SendFuture::new(async {
+            self.fetch_one(
+                "INSERT INTO spans
```

Comment:
are we going to have different sql queries per store? i thought wed have the actual methods be defaulted in the trait to just call the `.fetch_one` and `.fetch_all` (which are the only ones actually implemented per-store)

Reply:
Yeah this version is by no means final (error handling is also not really there. 

It would be nice if we are able to implement it through the trait and I think this is fine for now. But the only things that we need to think about is are 1) different sql dialects (ie. `?` vs `$1`) and 2) how will we support non sql environments? Sure those can just override everything, I guess.